sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -186,16 +207,60 @@ public Routing getRouting(WhereClause whereClause, @Nullable String preference)
         } catch (IndexMissingException e) {
             return new Routing();
         }
+
+        final List<ShardId> missingShards = new ArrayList<>(0);
         ShardRouting shardRouting;
         for (ShardIterator shardIterator : shardIterators) {
             shardRouting = shardIterator.nextOrNull();
-            if (shardRouting != null && shardRouting.active()) {
-                processShardRouting(locations, shardRouting);
+            if (shardRouting != null) {
+                if (shardRouting.active()) {
+                    processShardRouting(locations, shardRouting);
+                } else {
+                    missingShards.add(shardIterator.shardId());
+                }
             } else {
                 throw new UnavailableShardsException(shardIterator.shardId());
             }
         }
-        return new Routing(locations);
+        if (missingShards.isEmpty()) {
+            return new Routing(locations);
+        } else {
+            final SettableFuture<Routing> futureRouting = SettableFuture.create();
+            observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    Thread thread = new Thread(new Runnable() {
```

Comment:
can you ensure that this is only called once (quasi recursive)? 
only one thread started? 
maybe check the future beforehand.
deregister this listener when done?